[ACIX-883] Auto-retry e2e tests in CI

Ishirui opened the pull request:
Note: this PR depends on #38227

<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Following-up on ACIX-859 and #36422, try a new method for automatically rerunning failing e2e tests. This new method does not re-run tests that are already known as flaky and whose failure is accepted, contrary to what the previous method would do.

### Motivation

Increase CI reliability and diminish the impact (lost time) of flaky (but not yet marked as flaky) tests

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

TBD

### Possible Drawbacks / Trade-offs

Increased pipeline runtime + lost visibility on which tests are failing/flaky

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Ishirui committed:
feat(new_e2e_tests): Get non-known-flaky e2e test failures

Ishirui committed:
feat(new_e2e_tests): Logic for selectively retrying failed tests

refactor(e2e): Use `os.path.relpath` instead of `str.partition`

fix(e2e): Fix regex for selecting specific tests

The regex we are constructing in order to retry a specific set of failed tests was not working as expected without the non-capturing group. Specifically, it was selecting way too many tests, including some that were not even requested in the beginning (due to similar names) !

Ishirui reviewed (commented):
Todo:
- Default retries to 0
- If retry logic enabled, pass "nth retry" and "max retries" to [`TearDownSuite`](https://github.com/DataDog/datadog-agent/blob/161f8c67d580706dfcc2ecca4a84941304ed3750/test/new-e2e/pkg/e2e/suite.go#L649) method to prevent tearing down the infra on test failure if there are going to be further retries
> Try to make it so that if we don't pass any `--max-retries` flag, the behavior does not change at all 
- In order to do pass these env vars, add it [here](https://github.com/DataDog/datadog-agent/blob/91f63ee656e33684f62168757ff9c1fe14f037cb/test/new-e2e/pkg/runner/parameters/store_env.go#L16) so it gets loaded as a global go var

Ishirui commented on .gitlab/e2e_install_packages/common.yml:
```diff
@@ -32,7 +32,7 @@
   script:
     - DATADOG_AGENT_API_KEY=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $INSTALL_SCRIPT_API_KEY_ORG2 token ) || exit $?; export DATADOG_AGENT_API_KEY
-    - dda inv -- -e new-e2e-tests.run --targets $TARGETS --junit-tar "junit-${CI_JOB_ID}.tgz" ${EXTRA_PARAMS} --src-agent-version $START_MAJOR_VERSION --dest-agent-version $END_MAJOR_VERSION --test-washer
+    - dda inv -- -e new-e2e-tests.run --targets $TARGETS --max-retries 3 --junit-tar "junit-${CI_JOB_ID}.tgz" ${EXTRA_PARAMS} --src-agent-version $START_MAJOR_VERSION --dest-agent-version $END_MAJOR_VERSION --test-washer
```
I added retries to all usages of the `new-e2e-tests.run` task, but maybe I should only add it to the one in the main `e2e.yml` file ? Not sure if we should enable retries these other instances of e2e tests.

Ishirui commented on tasks/new_e2e_tests.py:
```diff
@@ -429,6 +465,28 @@ def write_result_to_log_files(logs_per_test, log_folder, test_depth=1):
             f.write("".join(logs))
 
 
+def _create_test_selection_regex(test_names: list[str]) -> str:
```
Not sure if this was the best place to put this, maybe I should move it to a directory under `libs/` or something ?

Ishirui replied:
Same comment now for the `filter_only_leaf_tests` function

KevinFairise2 replied:
Yes I think we can move them in `libs/`. Maybe `libs/testing` would be a goo d place

Ishirui commented on test/new-e2e/pkg/e2e/suite.go:
```diff
@@ -664,6 +665,26 @@ func (bs *BaseSuite[Env]) TearDownSuite() {
+		remainingRetries = 0
+	}
+
+	// Note: this check might sometimes skip stack deletion even if it would be needed
+	// For example if the only tests that failed were known flaky tests and thus not retried.
+	// This is not too big a deal as the stack is cleaned up manually in the invoke task handling retries.
```
Is this too dirty ? I wasn't sure if it was better to recompute all the logic for determining whether we should retry in here or not, it seems very cumbersome for not that much protection

Ishirui replied:
Not relevant anymore

KevinFairise2 commented on tasks/new_e2e_tests.py:
```diff
@@ -195,18 +197,54 @@ def run(
+
+    # In some cases the retry logic might leave resources undestroyed - make sure they are cleaned up
+    if max_retries > 0:
+        _clean_stacks(ctx, skip_destroy=not keep_stacks)
```
Do you think we can find a way to make sure that this call ignore the stack we explicitly marked to be kept?

KevinFairise2 commented on tasks/new_e2e_tests.py:
```diff
@@ -195,18 +197,54 @@ def run(
+            break
+
+    # In some cases the retry logic might leave resources undestroyed - make sure they are cleaned up
+    if max_retries > 0:
```
Let's use `max_retries > 0 and not keep_stacks`.
Calling `_clean_stacks` with `skip_destroy` will not destroy the stack but still remove it from the remote state, which will make us unable to destroy it properly later on

Ishirui replied:
Right - so we should always set `skip_destroy = False` right ?

Ishirui committed:
feat(e2e): Add `teardownOnly` parameter for running the suite teardown function independently of any tests

Ishirui committed:
fix(e2e): Avoid deleting infra if test will be retried + ensure everything is cleaned up properly if necessary

Compared to the previous solution, this method makes use of the "real" `TearDownSuite` function instead of blindly cleaning everything up. This makes sure we do not clean up infra for tests marked with `skipDeleteOnFailure` and similar conditions.

feat(e2e): Add extra log message for showing skipped retries

Ishirui committed:
feat(e2e): Filter out non-leaf jobs to make sure we avoid retrying useless tests

Ishirui commented on .gitlab/source_test/linux.yml:
```diff
@@ -196,7 +196,7 @@ new-e2e-unit-tests:
   script:
     - dda inv -- -e gitlab.generate-ci-visibility-links --output=$EXTERNAL_LINKS_PATH
-    - dda inv -- -e new-e2e-tests.run --targets ./pkg/utils --junit-tar junit-${CI_JOB_ID}.tgz ${EXTRA_PARAMS}
+    - dda inv -- -e new-e2e-tests.run --targets ./pkg/utils --max-retries 3 --junit-tar junit-${CI_JOB_ID}.tgz ${EXTRA_PARAMS}
```
Done :+1:

Ishirui committed:
fix(e2e): Make sure running with `--max-retries=0` has 0 behavioral impact

Ishirui committed:
feat(e2e): Run `install_packages` e2e tests with retries

KevinFairise2 committed:
Call AfterTest/BeforeTest in containers suite

(cherry picked from commit a7442377e322c0b3eb77628ed662e7d0952bb639)

Ishirui commented on tasks/new_e2e_tests.py:
```diff
@@ -362,18 +365,98 @@ def run(
+            test_profiler=None,
+        )
 
     success = process_test_result(test_res, junit_tar, AgentFlavor.base, test_washer)
```
TODO: Change how this works. Due to our retries, the json file gets constantly overwritten, so this does not hold any more meaning anymore - this especially bad when there is a `teardown_only` run, as that will _also_ overwrite the json file.
The best solution is probably to make sure that the json file gets _updated_ on each retry, instead of overwritten - except for the teardown, in which case we just write to `/dev/null`.

Ishirui commented:
/trigger-ci

Ishirui committed:
fix(e2e): Don't overwrite the result json file in a loop

Ishirui committed:
feat(e2e): Run `new-e2e-installer` jobs with retries

Ishirui committed:
feat(e2e): Add multiple junit files in uploaded `junit_tar`, one per attempt

Ishirui committed:
refactor(e2e): Move helper functions to `libs/testing/e2e.py`
